### ISSUE TITLE
Suddenly receiving error "chrome failed to start" and ALL requests fail after that

### ISSUE THREAD

**Issue by spewu** [gotenberg/gotenberg#633]
I guys

We are about to switch all of our PDF renderings to gotenberg ... but we just experienced a problem 😅

**Context**
We run the gotenberg container inside a Kubernetes cluster on Azure Kubernetes Service

**Problem**
Today, the embedded chrome app suddenly started failing. Here is an extract of the logs, at the time the failure happened:

```
{
    "level": "error",
    "ts": 1681978181.8257205,
    "logger": "api",
    "msg": "convert HTML to PDF: convert to PDF: chromium PDF: chrome failed to start:\n[0420/080941.795833:ERROR:platform_thread_posix.cc(157)] pthread_create: Resource temporarily unavailable (11)\n[0420/080941.796158:ERROR:platform_thread_posix.cc(157)] pthread_create: Resource temporarily unavailable (11)\n[0420/080941.796266:ERROR:platform_thread_posix.cc(157)] pthread_create: Resource temporarily unavailable (11)\n[0420/080941.796275:FATAL:browser_task_executor.cc(374)] Failed to start BrowserThread:IO\n[0420/080941.796346:ERROR:platform_thread_posix.cc(157)] pthread_create: Resource temporarily unavailable (11)\n[0420/080941.801669:ERROR:directory_reader_posix.cc(42)] opendir /tmp/Crashpad/attachments/17d2a248-cad8-4fe1-b768-05b484d094e7: No such file or directory (2)\n[0420/080941.809197:ERROR:http_transport_libcurl.cc(470)] curl_easy_perform: Couldn't resolve host name (6)\n",
    "trace": "e37d34ce-52ac-4dbe-adc3-2dd4233d7ef4",
    "remote_ip": "10.244.8.11",
    "host": "likvido-gotenberg.pdf.svc.cluster.local",
    "uri": "/forms/chromium/convert/html",
    "method": "POST",
    "path": "/forms/chromium/convert/html",
    "referer": "",
    "user_agent": "",
    "status": 500,
    "latency": 81477599,
    "latency_human": "81.477599ms",
    "bytes_in": 14582,
    "bytes_out": 21
}
```

After this first failure, all requests failed with the following error:

```
{
    "level": "error",
    "ts": 1681978184.03036,
    "logger": "api",
    "msg": "convert HTML to PDF: convert to PDF: chromium PDF: chrome failed to start:\n[0420/080943.991640:ERROR:platform_thread_posix.cc(157)] pthread_create: Resource temporarily unavailable (11)\n",
    "trace": "c4fed7b4-ecbd-457e-9611-02c9118c5563",
    "remote_ip": "10.244.8.11",
    "host": "likvido-gotenberg.pdf.svc.cluster.local",
    "uri": "/forms/chromium/convert/html",
    "method": "POST",
    "path": "/forms/chromium/convert/html",
    "referer": "",
    "user_agent": "",
    "status": 500,
    "latency": 62128927,
    "latency_human": "62.128927ms",
    "bytes_in": 43474,
    "bytes_out": 21
}
```

Full logs:
[likvido-gotenberg-7bdc9b4776-77q4n.log](https://github.com/gotenberg/gotenberg/files/11284713/likvido-gotenberg-7bdc9b4776-77q4n.log)

I noticed in our logs that the service making calls to the gotenberg API started reporting many errors, and when I looked at the logs from the gotenberg container, I decided to restart the "deployment" in Kubernetes, which means Kubernetes spins up another instance of gotenberg, and kills the old instance. Once the new instance came up, it was working fine again.

So, basically, I have two issues here:

1. The actual problem that happened with Chrome suddenly not being able to start
2. The fact that after Chrome was unable to start, ALL requests failed, but the gotenberg container did not crash or exit, which means Kubernetes was not aware of any issues, and it would not restart the deployment. If something like this happens, it would be nice if the container would crash or exit, such that Kubernetes would start a new instance automatically.

Thanks a lot!

**Comment by gulien** [comment]
Hello @spewu,

Thanks for the details. A new health check that verifies if the last X errors are of `chrome failed to start` kind could work.

**Comment by spewu** [comment]
> Hello @spewu,
> 
> Thanks for the details. A new health check that verifies if the last X errors are of `chrome failed to start` kind could work.

yea, that would be cool. Is that something I can configure already? - or would it require a new release with some changes?

**Comment by gulien** [comment]
No, someone has to write it 😅

**Comment by spewu** [comment]
> No, someone has to write it 😅

OK, fair enough 😅

Unfortunately I don't know Go, so I am probably not a very good candidate for writing this, but do you have any guidance on how to go about adding such a health check?

**Comment by gulien** [comment]
Hello @spewu,

Could you test that the new health check is working as expected with the Docker image `gulnap/gotenberg:chromium-hc` 
?

It comes with the following new flag: 

```
chromium-failed-starts-threshold int Set the number of failed starts after which the module is considered unhealthy - 0 means ignore (default 5)
```

**Comment by spewu** [comment]
> 

Wow, that is amazing!
Thank you very much for acting so quickly on this. I had not expected that, so I already added a cron job that restarts the gotenberg deployment every hour 😅

I will need to build a new image, based on your image, that contains our fonts, and then I will deploy it to our cluster, and remove the new cron job.
Unfortunately, I don't know how to trigger the problem myself, so I will have to monitor the cluster while we do this testing, to see when it happens again, that it will recover automatically from this.

When `the module becomes unhealthy`, what happens exactly? - will the container exit?

**Comment by gulien** [comment]
It means the the `/health` route now returns a 503 ^^

**Comment by spewu** [comment]
> It means the `/health` route now returns a 503 ^^

Ahh, super cool. So I can use a `livenessProbe` to restart the deployment if it fails. Thanks a lot @gulien!

I will deploy the test using your image now, and get back to you here once I have any results.

**Comment by spewu** [comment]
@gulien I haven't forgotten to get back to you, but the weird thing is that I have not seen the problem since I switched to your `gulnap/gotenberg:chromium-hc` image, so I am still waiting for it to happen again.

**Comment by gulien** [comment]
Hello @spewu! No problem. Note there is also a new version of Google Chrome, it might be more stable.

**Comment by spewu** [comment]
> Hello @spewu! No problem. Note there is also a new version of Google Chrome, it might be more stable.

Ahh, ok - that is interesting, so maybe they actually fixed the underlying issue. Anyway, I will keep monitoring and report back here if it happens again 🙂

**Comment by spewu** [comment]
@gulien OK, I am an idiot - I didn't know Kubernetes actually reuses the same pod when the liveness probe fails. I just looked back over the logs from the container in the pod, and I see it does restart the container occasionally. So I guess this means the liveness probe works and solves my problem.

Thanks a lot!

**Comment by gulien** [comment]
Thanks for the feedback @spewu 👍
